### PR TITLE
Map editor menu improvement and refactoring

### PR DESCRIPTION
**Pull request by Matroftt**
**Describe what the proposed change does**
- Players can clear field, without leaving map editor and opening the map editor again
- Added options to map editor
- Players can overlay on map like v0.30 map editor (flags, parcel, positions) in options

_Edit: Added by tobylane:_
- _Refector common parts of the menu_
- _Add music controls to the map editor_

### CODE REVIEW

**Review by TheCycoONE — CHANGES_REQUESTED** [review]
It would be cleaner to elevate the functions out of the game menu and reuse them here rather than copy/pasting the code.

I'm in favour of adding the options. The PR needs some refactoring.

**Comment by lewri** [comment]
Just to say regardless thanks Matvey for looking at this! :)

File: CorsixTH/Lua/dialogs/menu.lua [hunk 1]
```diff
@@ -545,11 +545,94 @@ end
 function UIMenuBar:makeMapeditorMenu(app)
   local menu = UIMenu()
   local hotkeys = app.hotkeys
-
   menu:appendItem(_S.menu_file.load:format(hotkey_value_label("ingame_loadMenu", hotkeys)), function() self.ui:addWindow(UILoadMap(self.ui, "map")) end)
     :appendItem(_S.menu_file.save:format(hotkey_value_label("ingame_saveMenu", hotkeys)), function() self.ui:addWindow(UISaveMap(self.ui)) end)
+	:appendItem(_S.menu_file.restart:format(hotkey_value_label("ingame_restartLevel", hotkeys)), function() app:mapEdit() end)
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:550]: Indent slightly off with lines above and below here

File: CorsixTH/Lua/dialogs/menu.lua [hunk 2]
```diff
@@ -545,11 +545,94 @@ end
...
     :appendItem(_S.menu_file.quit:format(hotkey_value_label("ingame_quitLevel", hotkeys)), function() self.ui:quit() end)
   self:addMenu(_S.menu.file, menu)
+  local options = UIMenu()
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:553]: This is already declared at line 546 (`local menu = UIMenu()`).

File: CorsixTH/Lua/dialogs/menu.lua [hunk 3]
```diff
@@ -545,11 +545,94 @@ end
...
+  if app.audio.has_bg_music then
+    local function vol(level, setting)
+      if setting == "music" then
+        return level == app.config.music_volume,
+          function()
+            app.audio:setBackgroundVolume(level)
+          end,
+          ""
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:561]: This might be correct, but what are the "" for?

File: CorsixTH/Lua/dialogs/menu.lua [hunk 4]
```diff
@@ -545,11 +545,94 @@ end
...
+    end,
+    nil,
+    function()
+      return app.config.play_music
+    end)
+
+  options
+    :appendMenu(_S.menu_options.sound_vol,         appendVolume("sound"))
+    :appendMenu(_S.menu_options.music_vol,         appendVolume("music"))
+    :appendItem(_S.menu_options.jukebox:format(hotkey_value_label("ingame_jukebox", hotkeys)), function() self.ui:addWindow(UIJukebox(app)) end)
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:609]:
```suggestion
  options:appendMenu(_S.menu_options.sound_vol, appendVolume("sound"))
    :appendMenu(_S.menu_options.music_vol, appendVolume("music"))
    :appendItem(_S.menu_options.jukebox:format(hotkey_value_label("ingame_jukebox", hotkeys)), function() self.ui:addWindow(UIJukebox(app)) end)
```

Sorry, never been keen on the excessive align for parts of the function params.

File: CorsixTH/Lua/dialogs/menu.lua [hunk 5]
```diff
@@ -545,11 +545,94 @@ end
...
   local menu = UIMenu()
   local hotkeys = app.hotkeys
-
   menu:appendItem(_S.menu_file.load:format(hotkey_value_label("ingame_loadMenu", hotkeys)), function() self.ui:addWindow(UILoadMap(self.ui, "map")) end)
     :appendItem(_S.menu_file.save:format(hotkey_value_label("ingame_saveMenu", hotkeys)), function() self.ui:addWindow(UISaveMap(self.ui)) end)
+	:appendItem(_S.menu_file.restart:format(hotkey_value_label("ingame_restartLevel", hotkeys)), function() app:mapEdit() end)
     :appendItem(_S.menu_file.quit:format(hotkey_value_label("ingame_quitLevel", hotkeys)), function() self.ui:quit() end)
   self:addMenu(_S.menu.file, menu)
+  local options = UIMenu()
+  if app.audio.has_bg_music then
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:554]: I can't easily see what line you `end` this if statement :(

File: CorsixTH/Lua/dialogs/menu.lua [hunk 6]
```diff
@@ -545,11 +545,94 @@ end
...
+        app.map:clearDebugText()
+      end
+    end
+end
+    options:appendMenu(_S.menu_debug.map_overlay,        UIMenu()
+        :appendCheckItem(_S.menu_debug_overlay.none,         true, overlay(), "")
+        :appendCheckItem(_S.menu_debug_overlay.flags,       false, overlay("flags"), "")
+        :appendCheckItem(_S.menu_debug_overlay.positions,   false, overlay("positions"), "")
+        :appendCheckItem(_S.menu_debug_overlay.parcel,      false, overlay("parcel"), "")
+	)
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:626]: Rid the alignment inside `appendMenu` and `appendCheckItem`

File: CorsixTH/Lua/dialogs/menu.lua [hunk 7]
```diff
@@ -545,11 +545,94 @@ end
...
+local function overlay(...)
+    local args = {n = select('#', ...), ...}
+    return function(item, m)
+      if args.n > 0 then
+        app.map:loadDebugText(unpack(args, 1, args.n))
+      else
+        app.map:clearDebugText()
+      end
+    end
+end
```
Comment (lewri) [inline CorsixTH/Lua/dialogs/menu.lua:620]: Ack :(

**Review by lewri — CHANGES_REQUESTED** [review]
Still work to be done in the alignment here. Unfortunately that's made it a bit difficult to read where something starts and ends (namely `if app.audio.has_bg_music then`)

File: CorsixTH/Lua/dialogs/menu.lua (hunk 4 above)
Comment (lewri) [reply CorsixTH/Lua/dialogs/menu.lua:609]: To note: 608 could be split across two lines but that's a smaller issue and one that can be left for now.

**Comment by TheCycoONE** [comment]
Do you mind if I fix this up @Matvey1705

**Comment by TheCycoONE** [comment]
No response, cannot merge as is.

**Comment by Matvey1705** [comment]
> Do you mind if I fix this up @Matvey1705


Its too, too late, if you still can do this please do it. Was checking notifications and found this. Sorry about that

**Comment by lewri** [comment]
Reopened at request from discord/matrix

**Comment by tobylane** [comment]
I've rebased, refactored, altered whitespace.

Should the restart map menu option have a confirm window?

**Comment by tobylane** [comment]
A whitespace fix, and a rebase. The map editor restart confirm window will come later.

Ready.

File: CorsixTH/Lua/dialogs/menu.lua [hunk 8]
```diff
@@ -540,6 +540,76 @@ local function hotkey_value_label(hotkey_name, hotkeys)
   return string.upper(array_join(hotkey_value, "+"))
 end
 
+local function audio_options(self, app, menu)
+  if not app.audio.has_bg_music then return end
+  local hotkeys = app.hotkeys
```
Comment (Alberth289346) [inline CorsixTH/Lua/dialogs/menu.lua:545]: - Add an empty line above this line. 

  The first line handles something else than this and the next lines. If you add an empty line between such blocks, you can find closely related code for some part more easily.

  Also, it helps in navigating. When looking for some specific part, you can start reading code and when you realize it's not what you are looking for, you can simply skip to code below the next empty line. You don't need to carefully keep reading to find the spot where the code starts doing something else.

  One step further, if you would start each such block with a single line of comment that describes what the code block is doing, you can navigate even quicker by reading the comment line, decide if the code block is what you're looking for, and if not directly skip to the next block with the next comment line above it.

- You don't seem to need `hotkeys` until further down in the function, so move that line closer to first use.

**Comment by lewri** [comment]
@tobylane can you just re-summarise what this PR is now doing as I guess its nature has changed.

You should also be able to add an 'Edit' note to the OP's post at the bottom with this.

>Should the restart mapeditor menu option have a confirm window?
Although you're now doing this later, yes it definitely should.

**Comment by lewri** [comment]
Are we good to merge this? I see no issue.

@tobylane does this need a rebase first? I don't think it does...

**Pull request merged**